File: web/src/lib/taskResume.ts
import type { VideoStatus } from '../types'

/** 失败任务是否可从笔记阶段继续（与后端 resume_available 一致，前端兜底判断） */
export function canResumeFromNotes(status: VideoStatus | null | undefined): boolean {
  if (!status || status.status !== 'error') return false
  if (status.resume_available) return true
  const segs = status.transcript_segments
  const text = (status.transcript || '').trim()
  if (!segs?.length || text.length < 50) return false
  if (status.error_message) return true
  const markers = ['处理已中断', '处理超时', '请重新上传', 'DeepSeek API', '笔记生成超过']
  return !markers.some((m) => text.startsWith(m) || text.includes(m))
}

export function formatTaskError(status: VideoStatus | null | undefined): string {
  const msg =
    status?.error_message?.trim() ||
    (status?.status === 'error' && !canResumeFromNotes(status)
      ? status.transcript?.trim()
      : '') ||
    '视频处理失败'
  return msg.length > 300 ? `${msg.slice(0, 300)}…` : msg
}

File: backend/core/task_resume.py
"""失败任务断点续跑：判断能否跳过 Whisper 仅重试笔记。"""

from __future__ import annotations

import json

_ERROR_MARKERS = (
    "处理已中断",
    "处理超时",
    "请重新上传",
    "DeepSeek API",
    "笔记生成超过",
    "转写进程",
    "转写输出",
    "无法连接 DeepSeek",
    "处理失败",
)


def _parse_segments(raw) -> list:
    if not raw:
        return []
    if isinstance(raw, list):
        return raw
    if isinstance(raw, str):
        try:
            parsed = json.loads(raw)
            return parsed if isinstance(parsed, list) else []
        except json.JSONDecodeError:
            return []
    return []


def video_has_resumable_transcript(video: dict) -> bool:
    """转写已落库且文本不像错误文案时，可从笔记阶段继续。"""
    segments = _parse_segments(video.get("transcript_segments"))
    if not segments:
        return False
    text = (video.get("transcript") or "").strip()
    if len(text) < 50:
        return False
    if (video.get("error_message") or "").strip():
        return True
    if video.get("status") != "error":
        return False
    head = text[:160]
    if any(m in head for m in _ERROR_MARKERS):
        return False
    return True


def video_error_message(video: dict) -> str | None:
    msg = (video.get("error_message") or "").strip()
    if msg:
        return msg
    if video.get("status") != "error":
        return None
    if video_has_resumable_transcript(video):
        return None
    t = (video.get("transcript") or "").strip()
    return t or None
